### DISCUSSION TITLE
Default value and async select options

### DISCUSSION THREAD

**Discussion by michalbundyra** [react-hook-form/react-hook-form#6086] [Q&A] [answered]
Recently I just noticed that when I am using select with options loaded from the server the first loaded option is selected, not the one set in `defaultValues` on `useForm` (my option with empty value is disabled, so once value is selected can't be unselected).

I've created simple example:

https://codesandbox.io/s/eloquent-bose-z95uu?file=/src/App.js

Please notice, that default value is user_id = 3, but when the form is loaded it will be user_id=1 selected.

When you submit the form without touching the select it will submit user_id=3, but if you touch the select, without changing anything it will change to user_id=1.

I found somewhere a bit hacky solution (imo) to solve that issue - to set the default values after loading, like that:

```javascript
reset({ ...getValues, 'user_id': 3 })
```

so I believe in the example above I would need to do:

```javascript
useEffect(() => {
  if (users.length === 0) {
    return
  }

  reset({ ...getValues(), 'user_id': 3 })
}, [reset, getValues, users])
```

Is there any better option to do that?
I've tried also with react-select and AsyncSelect there, but it's even more complicated, as the defaultValue must be an object.

Thanks for any advice!

**Comment by bluebill1049** [comment, marked as answer]
can you test the latest? I think it's fixed.

**Reply by michalbundyra** [reply]
@bluebill1049 do you mean 7.12.1 release? 

Just checked on codesandbox and it looks the same - still wrong.
Does it work for you as expected on the codesandbox after using 7.12.1?

**Reply by bluebill1049** [reply]
I think that's expected behavior.

- `defaultvalue` is 3 which is the user id
- after loaded, you are only reset the users, this doesn't mean you are updating the input value either it should read the first select option (hook form is uncontrolled by default)
- after you touch the input, the select option is getting updated.

perhaps you want to invoke the `setValue` after inputs are loaded.

**Reply by michalbundyra** [reply]
So I would need something like that:

```javascript
  React.useEffect(() => {
    if (users.length > 0) {
      setValue('user_id', 3)
    }
  }, [users, setValue])
```

right?

I just thought that there might be a better option, if not, it's fine, I'll look for something I can do on my side.

If this is the only solution (and the recommended one) what about adding it to the examples? Thanks

**Reply by michalbundyra** [reply]
@bluebill1049 per your second point I've tried to use controlled field instead:

```javascript
<Controller
  name="created_by"
  control={control}
  rules={{
    valueAsNumber: true
  }}
  render={({field}) => (
    <select {...field}>
      <option disabled>Please select...</option>
      {users.map((user) => (
         <option key={user.id} value={user.id}>
           {user.id} : {user.name}
         </option>
      ))}
    </select>
  )}
/>
```

and it works as expected, the correct value will be selected here once list of user is loaded, without using setValue or reset. What I have is just `defaultValues` in `useForm`.

**Reply by bluebill1049** [reply]
nice! that's what i was going to quickly try out.

**Reply by bluebill1049** [reply]
you can remove `valueAsNumber`.

**Reply by michalbundyra** [reply]
If I remove valueAsNumber values are send as strings - so it is needed in my case.

**Reply by bluebill1049** [reply]
`Controller` doesn't have `valueAsNumber` option. I know this is going to start another discussion on why it's not there in the first place, but `valueAs` was borrowed from native API, and I know it's going to cause inconvenience because the registered support that, but really I don't want to go into the `tranform` APIish stuff, it's going to create a lot of issues as I can see. it's better to let the controlled component handle the data transform, it's a pain to have another wrapper but if you are going to reuse such a component probably worth the effort.

**Reply by michalbundyra** [reply]
Not sure what you are saying here - it works now with `valueAsNumber`, but it shouldn't?

The thing is that I want to have a generic component and either I need to pass rules or additional flag (for example `isNumber`) to handle it different way in onChange handler, i.e.:

```javascript
onChange={(e) => {
  field.onChange(isNumber ? Number(e.currentTarget.value) : String(e.currentTarget.value))
})
```

Just to summarise - for async loading options I have to use controlled, field - right?

**Reply by bluebill1049** [reply]
>Just to summarise - for async loading options I have to use controlled, field - right? 

yes. 

> Not sure what you are saying here - it works now with valueAsNumber, but it shouldn't?

that's not supported API, well if it works so be it. `valueAs` is really one-way data transform. only transform what's in the input, not what's provided by the defaultValue or defaultValues.
